shadowsocks/cipher_list.go

```diff
@@ -16,23 +16,26 @@ package shadowsocks
 
 import (
 	"container/list"
+	"net"
 	"sync"
 
 	"github.com/shadowsocks/go-shadowsocks2/shadowaead"
 )
 
 // CipherEntry holds a Cipher with an identifier.
+// The public fields are constant, but lastAddress is mutable under cipherList.mu.
 type CipherEntry struct {
-	ID     string
-	Cipher shadowaead.Cipher
+	ID          string
+	Cipher      shadowaead.Cipher
+	lastAddress net.IP
 }
 
 // CipherList is a list of CipherEntry elements that allows for thread-safe snapshotting and
 // moving to front.
 type CipherList interface {
 	PushBack(id string, cipher shadowaead.Cipher) *list.Element
-	SafeSnapshot() []*list.Element
-	SafeMoveToFront(e *list.Element)
+	SafeSnapshot(addr net.IP) []*list.Element
```

Comment:
Let's rename to something like `SafeSnapshotForClientIP(clientIP)`

Reply:
Done